lib/ash_authentication/strategies/password/sign_in_preparation.ex

```diff
@@ -46,9 +46,23 @@ defmodule AshAuthentication.Strategy.Password.SignInPreparation do
              password,
              Map.get(record, strategy.hashed_password_field)
            ) do
-          token_type = query.context[:token_type] || :user
-
-          {:ok, [maybe_generate_token(token_type, record, strategy)]}
+          if user_confirmed_if_needed(record, strategy) do
+            token_type = query.context[:token_type] || :user
+
+            {:ok, [maybe_generate_token(token_type, record, strategy)]}
+          else
+            {:error,
+             AuthenticationFailed.exception(
```

Comment:
This should add a custom exception, like `AshAuthentication.Errors.CannotConfirmUnconfirmedUser`